### PR TITLE
Propose a workaround for SwiftPM's lack of recursive header search paths

### PR DESCRIPTION
**Pull request by tonyarnold**
## Description

This PR proposes a workaround for the issue described in #2140. 

## Related PRs or issues

- #2140

## Misc

Ideally the list of header search paths could be scanned from the filesystem, however I think there are potentially bigger reliability issues with doing that than with what I've proposed.

### CODE REVIEW

**Comment by AnastasiaKubova** [comment]
Hi! Thanks for your contribution PR! 
I will check it and back to you with the results.

File: Package@swift-5.3.swift [hunk 1]
```diff
@@ -35,11 +35,14 @@ let package = Package(
             name: "AppCenter",
             path: "AppCenter/AppCenter",
             exclude: ["Support"],
-            cSettings: [
-                .define("APP_CENTER_C_VERSION", to:"\"4.4.2\""),
-                .define("APP_CENTER_C_BUILD", to:"\"1\""),
-                .headerSearchPath("**"),
-            ],
+            cSettings: {
```
Comment (AnastasiaKubova) [inline Package@swift-5.3.swift:38]: Hi! I checked your changes, but it doesn't work for me.
When I'm trying to call the `swift test` command (from the feature description)  I get the next error:
```
error: manifest parse error(s):
Unknown error parsing manifest for appcenter-sdk-apple 
```

Comment (tonyarnold) [reply Package@swift-5.3.swift:38]: Oh, right, I need to update that. That was `swift test` in local projects that use AppCenter SDK. 

To test this for the project, you would need to specify `swift build`.

Comment (tonyarnold) [reply Package@swift-5.3.swift:82]: My apologies, there was an error in the committed code. 

Could you please try running:

```sh
swift build --product AppCenterAnalytics
swift build --product AppCenterCrashes
```

These should work. I'm looking to see if I can make plain `swift build` work, but there's iOS code mixed in via the AppCenterDistribute target that isn't being excluded from the Swift package properly.

Comment (AnastasiaKubova) [reply Package@swift-5.3.swift:82]: Thanks for the update! 
Now it works for me with the `swift build` command, but fails with the `swift test` with error:
```
fatal error: 'UIKit/UIScreen.h' file not found
#import <UIKit/UIScreen.h>
        ^~~~~~~~~~~~~~~~~~
```

We will discuss these changes with the team and come back with the results.

Comment (tonyarnold) [reply Package@swift-5.3.swift:82]: Thank you. I realise it's not an ideal solution to the problem, but AppCenter not working with SwiftPM outside of Xcode is an issue that's going to continue to grow as people migrate their projects to use more Swift packages.

**Comment by DmitriyKirakosyan** [comment]
Hi @tonyarnold, I tested your solution and it does work fine for `AppCenterAnalytics` and `AppCenterCrashes` products. It doesn't work for `AppCenterDistribute` though. It is not building because of the UIKit dep. Is there any reason for adding Distribute dependency for a Swift Package project? If there is no such need, I guess we can ignore the broken Distribute product.

As an alternative solution, I was trying to add header paths recursively using a helper function with FileManager etc, however the solution wasn't elegant, so specifying paths explicitly looks better.

There is an issue in your changes though. If you try to integrate AppCenter SDK via SPM from your branch and add all modules, it would not compile because you missed some header paths for Distribute module.
I think we need to add "**" to the header path list to make Xcode happy when integrating SDK to an iOS project.
Could you please fix that?

**Comment by tonyarnold** [comment]
Thanks, @DmitriyKirakosyan. I've added "**" as a header search path in 09e5806231aba77ce6533a0a4ac7c02a6a5968a2. 

> Is there any reason for adding Distribute dependency for a Swift Package project? 

Not for macOS projects - Distribute is iOS-only, right? It should work when targeting iOS. 

>As an alternative solution, I was trying to add header paths recursively using a helper function with FileManager etc, however the solution wasn't elegant, so specifying paths explicitly looks better.

I tried the same, but yeah, I found the same outcome - it got messy quickly. This isn't great, but perhaps someone from Microsoft could file an issue on https://bugs.swift.org/ about supporting recursive header search paths in Swift Package Manager? (it's going to hold more weight than if I did it!)

**Comment by DmitriyKirakosyan** [comment]
@tonyarnold Thanks for the update.

> Not for macOS projects - Distribute is iOS-only, right? It should work when targeting iOS.

Yes, but what about iOS package projects? Probably worth mentioning that AppCenterDistribute product is not building via swift cli due to its iOS dependency.

> This isn't great, but perhaps someone from Microsoft could file an issue on https://bugs.swift.org/ about supporting recursive header search paths in Swift Package Manager?

I'm actually surprised no one has filed such feature request yet. I created one here: https://bugs.swift.org/browse/SR-15851.
Hopefully they fix it and we will be able to get rid of this workaround.

**Comment by tonyarnold** [comment]
I've updated the changelog to reflect that only the Analytics and Crashes targets will build via `swift test`, etc.

**Review by DmitriyKirakosyan — APPROVED** [review]

**Comment by DmitriyKirakosyan** [comment]
@tonyarnold merged. Thank you for your contribution!

**Comment by tonyarnold** [comment]
Thanks for accepting the patch! Hopefully Apple addresses the recursive header search limitations in SwiftPM, so that we can drop this workaround. 

Is there a release of this project planned anytime soon? It would be good to know when I can stop using my workarounds locally to get this working.

**Comment by DmitriyKirakosyan** [comment]
Next release is planned at the end of this quarter. Meantime, you can just use `develop` branch.

**Pull request merged**